Add Svnapot and Svpbmt extensions

ved-rivos opened the pull request.

jrtc27 reviewed (changes requested):
extPte is used by sail-cheri-riscv. Being an ext* thing it's not to be used in the standard model lest you break such downstream use cases.

ved-rivos commented:
> extPte is used by sail-cheri-riscv. Being an ext* thing it's not to be used in the standard model lest you break such downstream use cases.

The extPte is a misnomer in the current model as being something to be used for non-standard extensions. These bits are reserved for standard extensions. CHERI use of these will be a non-conforming extension which is fine and CHERI should get added accordingly.

jrtc27 commented:
> > extPte is used by sail-cheri-riscv. Being an ext* thing it's not to be used in the standard model lest you break such downstream use cases.
> 
> The extPte is a misnomer in the current model as being something to be used for non-standard extensions. These bits are reserved for standard extensions. CHERI use of these will be a non-conforming extension which is fine and CHERI should get added accordingly. Non-tree non-standard extensions should not be using bits reserved for future standard use. I think CHERI should define a new virtual memory system mode where it can put new non-RISC-V interpretations on these bits.

It's ext not stdExt, and there was no non-reserved way to extend the page table format in a custom extension at the time it was created. I am just asking you to not break existing uses of the model, one that has been around for many years and was developed at the same time and influenced the design of the base Sail model.

ved-rivos commented:
> > > extPte is used by sail-cheri-riscv. Being an ext* thing it's not to be used in the standard model lest you break such downstream use cases.
> > 
> > 
> > The extPte is a misnomer in the current model as being something to be used for non-standard extensions. These bits are reserved for standard extensions. CHERI use of these will be a non-conforming extension which is fine and CHERI should get added accordingly. Non-tree non-standard extensions should not be using bits reserved for future standard use. I think CHERI should define a new virtual memory system mode where it can put new non-RISC-V interpretations on these bits.
> 
> It's ext not stdExt, and there was no non-reserved way to extend the page table format in a custom extension at the time it was created. I am just asking you to not break existing uses of the model, one that has been around for many years and was developed at the same time and influenced the design of the base Sail model.

If a custom extension needs to redefine PTE layout then the right way would be to use a custom `satp.MODE`. The PTE layouts defined in RISC-V apply to the standard modes defined by RISC-V for `satp.MODE`. The model can then qualify these checks by the `satp.MODE` and the custom extension can provide its own implementation for validating PTEs. It would be wrong to hold back Svnapot and Svpbmt standard extensions from the golden reference model to avoid breaking an unrelated out of tree usage.

jrtc27 commented:
> > > > extPte is used by sail-cheri-riscv. Being an ext* thing it's not to be used in the standard model lest you break such downstream use cases.
> > > 
> > > 
> > > The extPte is a misnomer in the current model as being something to be used for non-standard extensions. These bits are reserved for standard extensions. CHERI use of these will be a non-conforming extension which is fine and CHERI should get added accordingly. Non-tree non-standard extensions should not be using bits reserved for future standard use. I think CHERI should define a new virtual memory system mode where it can put new non-RISC-V interpretations on these bits.
> > 
> > 
> > It's ext not stdExt, and there was no non-reserved way to extend the page table format in a custom extension at the time it was created. I am just asking you to not break existing uses of the model, one that has been around for many years and was developed at the same time and influenced the design of the base Sail model.
> 
> If a custom extension needs to redefine PTE layout then the right way would be to use a custom `satp.MODE`. The PTE layouts defined in RISC-V apply to the standard modes defined by RISC-V for `satp.MODE`. The model can then qualify these checks by the `satp.MODE` and the custom extension can provide its own implementation for validating PTEs. It would be wrong to hold back Svnapot and Svpbmt standard extensions from the golden reference model to avoid breaking an unrelated out of tree usage.

CHERI-RISC-V predates all of that. So no, we're not changing what was specified in the past. Ratified CHERI-RISC-V is on the way and will be standard, whatever that means for the design, but we are not changing our current specification. You may not care about CHERI, but CHERI-RISC-V is a big reason the base model exists at all. I'm not saying not to add Svpbmt and Svnapot support, I'm saying to add it in such a way that the existing CHERI-RISC-V model can opt out of it. Otherwise you are breaking a real-world use case that the model has always been intended to support from the very first moment it was written.

ved-rivos commented:
> CHERI-RISC-V predates all of that. So no, we're not changing what was specified in the past. Ratified CHERI-RISC-V is on the way and will be standard, whatever that means for the design, but we are not changing our current specification. You may not care about CHERI, but CHERI-RISC-V is a big reason the base model exists at all. I'm not saying not to add Svpbmt and Svnapot support, I'm saying to add it in such a way that the existing CHERI-RISC-V model can opt out of it. Otherwise you are breaking a real-world use case that the model has always been intended to support from the very first moment it was written.

It likely does predate but this repo is the RISC-V reference model and that is the primary goal. Serving non tree extensions would be secondary goal. As to how to make an out of tree usage that conflicts with the base RISC-V architecture work I lack enough known about CHERI and what its specification is. So suggest you create a PR with a suggestion on how that can be made to work and we can review it here.

Alasdair commented:
@jrtc27 I agree we should not break CHERI, do you have a suggestion for how we can avoid this? It seems to me like we could have something like:
```
match ext_isInvalidPTE(p, ext) {
  Some(is_valid) => is_valid,
  None() => <standard logic> 
}
```
The base model would then have an implementation of ext_isInvalidPTE that always returns `None()`, which could be overridden by CHERI.

The model uses function prefixed `ext_` has 'hook' points for out of tree uses to adapt the behavior of the model. I think using `EXT` or `ext` as prefix like in `EXT_PTE_Bits` is probably confusing given this.

jrtc27 commented:
> @jrtc27 I agree we should not break CHERI, do you have a suggestion for how we can avoid this? It seems to me like we could have something like:
> 
> ```
> match ext_isInvalidPTE(p, ext) {
>   Some(is_valid) => is_valid,
>   None() => <standard logic> 
> }
> ```
> 
> The base model would then have an implementation of ext_isInvalidPTE that always returns `None()`, which could be overridden by CHERI.
> 
> The model uses function prefixed `ext_` has 'hook' points for out of tree uses to adapt the behavior of the model. I think using `EXT` or `ext` as prefix like in `EXT_PTE_Bits` is probably confusing given this.

That seems sufficient and pretty straightforward.

ved-rivos commented:
> The base model would then have an implementation of ext_isInvalidPTE that always returns `None()`, which could be overridden by CHERI.

@Alasdair The use of those bits in walk39/walk48 will also need to be qualified by haveSvnapot() otherwise the walk if it finds the N bit set will interpret the PTE to be in Svnapot form. This should be a small update. I will push an update with this idea. The haveSvnapot() however is hard coded to true (like all other extensions). Should that be made a command line switch? Or do the out of tree uses have a way around this?

jrtc27 commented:
We don't, so that will need some way to turn it off without modifying non-ext files.

ved-rivos commented:
I will add a command line switch. Maybe its time to add an isa parser.

> I think using EXT or ext as prefix like in EXT_PTE_Bits is probably confusing given this.

`extPte` can be renamed as `pteStdExtAttribs` and `EXT_PTE_Bits` to `PTE_StdExt_Bits` (and likewise for other names like `ext_pte`, `default_sv32_ext_pte`, `pte.Ext()`. Does that look better?

jrtc27 commented:
> I will add a command line switch.

That doesn't really help out-of-tree extensions override it, because the command-line switch lives in the .c and .ml files, so you'd be requiring sail-cheri-riscv users to always disable Svnapot on the command-line, which is not a helpful interface. It needs to be disableable *statically at build time* without modifying non-ext files.

ved-rivos commented:
> > I will add a command line switch.
> 
> That doesn't really help out-of-tree extensions override it, because the command-line switch lives in the .c and .ml files, so you'd be requiring sail-cheri-riscv users to always disable Svnapot on the command-line, which is not a helpful interface. It needs to be disableable _statically at build time_ without modifying non-ext files.

I planned to have a command line switch to enable it and not disable. The current practise of statically enabling every extension in Sail RISC-V model is also a problem because people cannot use it without coded modification as a reference model. So this switch will be temporary and we should have a isa string parser so people can customize the reference to their DUT specific configuration.

jrtc27 commented:
> > > I will add a command line switch.
> > 
> > 
> > That doesn't really help out-of-tree extensions override it, because the command-line switch lives in the .c and .ml files, so you'd be requiring sail-cheri-riscv users to always disable Svnapot on the command-line, which is not a helpful interface. It needs to be disableable _statically at build time_ without modifying non-ext files.
> 
> I planned to have a command line switch to enable it and not disable. The current practice of statically enabling every extension in Sail RISC-V model is a problem because people cannot use it without code modification as a reference model. This switch will be temporary and we should have a ISA string parser so people can customize the reference to their DUT specific configuration without needing source changes.

Ok, off-by-default works.

arichardson commented on test/riscv-tests/rv64ssvnapot-p-napot.elf:
I think it's great that you've added a test for this change, but I don't think we should be checking in these huge binaries anymore. I'd prefer to either keep them external or use assembly.

ved-rivos replied:
I agree we should move to a better test framework. 

> I'd prefer to either keep them external or use assembly.

Please suggest how to integrated with CI while keeping them external or use assembly.

arichardson replied:
CI can always download pre-built test binaries from another source. Git is really not great when it comes to large files so I'd very much like the repository to remain a reasonable size. The existing checked-in binaries should probably also be deleted, but since they are already here there is not much we can do about them.

ved-rivos replied:
Sounds good. I have dropped the elf file from the PR.

ved-rivos commented:
Updated as follows:
1. Add --enable-svnapot and --enable-svpbmt command line switches. In tree usages will need to provide the switch to explicitly enable the RISC-V standard extensions. Out of tree usages will not have to do more work. 
2. Add `ext_isInvalidPTE` - returns invalid, valid, or standard. If custom extension check fails should return invalid. To skip standard extension checks returns valid. Default is to invoke standard extension checks. Renamed `ext_pte` to avoid `std_ext`.
3. Dropped the test while we figure out what the test framework for the model should be and how to integrate that into CI.

billmcspadden-riscv commented:
I'm working on a test framework as the current framework doesn't allow
for testing of things like command line switches.  I'll be presenting at the
tech-golden-model meeting on 2024-1-29.

Bill Mc.


On Mon, Jan 22, 2024 at 12:50 PM Ved Shanbhogue ***@***.***>
wrote:

> Updated as follows:
>
>    1. Add --enable-svnapot and --enable-svpbmt command line switches. In
>    tree usages will need to provide the switch to explicitly enable the RISC-V
>    standard extensions. Out of tree usages will not have to do more work.
>    2. Add ext_isInvalidPTE - returns invalid, valid, or standard. If
>    custom extension check fails should return invalid. To skip standard
>    extension checks returns valid. Default is to invoke standard extension
>    checks. Renamed ext_pte to avoid std_ext.
>    3. Dropped the test while we figure out what the test framework for
>    the model should be and how to integrate that into CI.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/393#issuecomment-1904783790>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXROLOHUFCIYEBYID24QIBDYP3GJZAVCNFSM6AAAAABCDRGRN2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBUG44DGNZZGA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Bill McSpadden
Formal Verification Engineer
RISC-V International
mobile:  503-807-9309

ved-rivos commented:
@jrtc27 - Did that address the change requested?

martinberger commented:
Quick question: given that purpose of Svpbmt is overriding of PMAs, how and where do you expect PMA checks to access the relevant bits 62–61 of a leaf page table entry?

ved-rivos commented:
@martinberger - for each load/store/instruction-fetch, we would first determine the PBMT from bits 62:61 of  the PTE entry that provides its system physical address. If the PBMT is IO or NC then those attributes apply. If PBMT=PMA, then the attributes associated with  the system physical address apply.

jrtc27 commented:
But where is that being captured in the model?

ved-rivos commented:
There are no PMAs modeled afaict.

martinberger commented:
The reason I asked is that the RISCV model will eventually get PMA support.  And I was wondering about the best way to get this PBMT bits to the PMA checks.

allenjbaum commented:
Ooh- Adding PMA support will be tricky, since there is no architectural
definition except that certain “categories” exist.
*Every* implementation is completely custom. This is possibly even harder
than WARL fields.

I did propose a syntax for describing PMAs, which could enable a paramete
rized version, but it is pretty simple,
and doesn't come close to being able to describe PMAs that are
programmable, or use any architectural state beyond address bits to make
decisions.


On Tuesday, March 12, 2024, Martin Berger ***@***.***> wrote:

> The reason I asked is that the RISCV model will eventually get PMA
> support. And I was wondering about the best way to get this PBMT bits to
> the PMA checks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/393#issuecomment-1992659202>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJS2YFID5NNAPUP7QITYX53Q3AVCNFSM6AAAAABCDRGRN2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJSGY2TSMRQGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

martinberger commented:
> Once PMA support is there this would be simple select between picking the PMA determined by SPA vs. override provided by PBMT.

@ved-rivos  If I understand the logic of your implementation correctly, you
suggest that the check of the PBMT bits happens in
`checkPTEPermission` in `riscv_pte.sail`? I'm not sure there is enough information available in `checkPTEPermission` to do this.

martinberger commented:
> @martinberger - the PMA checks should happen parallel to where the PMP checks happen currently - not in checkPTEPermission. The pipeline would look as follows: Load/Store -> [TLB/PageTableWalk] -> SPA + PBMT + size of access + type of access (read/write/implicit-read/implicit-write) -> [PMA + PMP] -> [MEMORY]

Yes, exactly. But right now the PBMT info is not returned to the caller, if I understand the logic of your implementation

ved-rivos commented:
> Yes, exactly. But right now the PBMT info is not returned to the caller, if I understand the logic of your implementation

When the PMA's are implemented then we would need the PBMT provided to caller. Right now there are no physical memory attributes in the model to provide to the caller.

martinberger commented:
> > Yes, exactly. But right now the PBMT info is not returned to the caller, if I understand the logic of your implementation
> 
> When the PMA's are implemented then we would need the PBMT provided to caller. Right now there are no physical memory attributes in the model to provide to the caller.

Yes, but a lot of RV companies will have their own local PMP/PMA already, and this will be even more in the future.  Why not make the PBMT information to the caller` already? Otherwise each company will cook up their own and upstreaming code becomes a tiny bit more burdensome?

ved-rivos commented:
Sure. That makes sense. Let me add that to the bundle.

martinberger commented:
> Sure. That makes sense. Let me add that to the bundle.

One possible way could be to change `TR_Result`  to something like:
```

type pbmt_bits = bits(2) // For Svpbmt extension                                                   

union TR_Result('paddr : Type, 'failure : Type) = {
  TR_Address : ('paddr, ext_ptw, pbmt_bits),
  TR_Failure : ('failure, ext_ptw)
}
```
(This is not the only way)

ved-rivos commented:
Is there a requirement that unions cannot be > 64 bits. I get this error with:
```
union TR_Result('paddr : Type, 'failure : Type) = {
  TR_Address : ('paddr, bits(10), ext_ptw),
  TR_Failure : ('failure, ext_ptw)
}
```


```
generated_definitions/c/riscv_model_RV64.c: In function ‘ztranslateAddr_priv’:
generated_definitions/c/riscv_model_RV64.c:61152:87: error: incompatible type for argument 1 of ‘convert_fbits_of_lbits’
61152 |         zstd_ext = CONVERT_OF(fbits, lbits)(zgaz32384.zTR_AddresszIUPTW_ErrorzIzKzCbzK.ztup1, true);
      |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
      |                                                                                       |
      |                                                                                       uint64_t {aka long unsigned int}
In file included from generated_definitions/c/riscv_model_RV64.c:1:
/home/ved/.opam/4.08.0/share/sail/lib/sail.h:311:32: note: expected ‘lbits’ but argument is of type ‘uint64_t’ {aka ‘long unsigned int’}
  311 | fbits CONVERT_OF(fbits, lbits)(const lbits, const bool);
      |                                ^~~~~~~~~~~
generated_definitions/c/riscv_model_RV64.c:61286:86: error: incompatible type for argument 1 of ‘convert_fbits_of_lbits’
61286 |         zuz3309 = CONVERT_OF(fbits, lbits)(zgaz32395.zTR_AddresszIUPTW_ErrorzIzKzCbzK.ztup1, true);
      |                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
      |                                                                                      |
      |                                                                                      uint64_t {aka long unsigned int}
In file included from generated_definitions/c/riscv_model_RV64.c:1:
/home/ved/.opam/4.08.0/share/sail/lib/sail.h:311:32: note: expected ‘lbits’ but argument is of type ‘uint64_t’ {aka ‘long unsigned int’}
  311 | fbits CONVERT_OF(fbits, lbits)(const lbits, const bool);

```

martinberger commented:
> Is there a requirement that unions cannot be > 64 bits. 

There is no such requirement. This might be a Sail compiler bug because this is the C-compiler complaining. Several such `incompatible type` problems were recently fixed. What version of the Sail compiler are you using? @Alasdair

ved-rivos commented:
I have - Sail 0.17.1 (sail @ opam-v2.1.2)

martinberger commented:
> I have - Sail 0.17.1 (sail @ opam-v2.1.2)

That's an old version and will fail on the problem I reported in https://github.com/rems-project/sail/issues/401. That problem was fixed last year. I suggest that you upgrade to a new version of the compiler. I don't know if the fix has already reached the official version that is distributed with `opam`.

jrtc27 commented:
There isn't a newer version tagged, and opam is just pointing at the 0.17.1 tag. Sounds like we need a new release?

martinberger commented:
> There isn't a newer version tagged, and opam is just pointing at the 0.17.1 tag. Sounds like we need a new release?

Yes, a new opam version would be good---Alasdair did a lot of work on the Compiler.

ved-rivos commented:
Updating the union declaration fixes the  error:
```
union TR_Result('paddr : Type, 'std_ext_bits : Type, 'failure : Type) = {
  TR_Address : ('paddr, 'std_ext_bits, ext_ptw),
  TR_Failure : ('failure, ext_ptw)
}
```

ved-rivos commented:
Updated to provide the PBMT bits (and other standard extension bits) along with the translation results to load/store/instruction-fetch.

ved-rivos commented:
@martinberger - does this look good? Are further changes needed?

martinberger commented:
> does this look good? Are further changes needed?

@ved-rivos LGTM. We should try and get it in.

martinberger commented:
@Timmmm  and I were thinking that it'd be better to convert the two PBMT bits directly at the source to a suitable symbolic representation and then send that to the PMA checks. That would give us stronger type-checks, and, at the same time, serve as documentation. How about this:

    enum PBMT = { PMA, NC, IO }

    mapping pbmt_enc : PBMT <-> bits(2) = {
      PMA <-> 0b00,
      NC  <-> 0b01,
      IO  <-> 0b10,
      // 0b11 is reserved.
    }

ved-rivos commented:
> @Timmmm and I were thinking that it'd be better to convert the two PBMT bits directly at the source to a suitable symbolic representation and then send that to the PMA checks. That would give us stronger type-checks, and, at the same time, serve as documentation. How about this:
> 
> ```
> enum PBMT = { PMA, NC, IO }
> 
> mapping pbmt_enc : PBMT <-> bits(2) = {
>   PMA <-> 0b00,
>   NC  <-> 0b01,
>   IO  <-> 0b10,
>   // 0b11 is reserved.
> }
> ```

The code presently provides all the extended PTE bits along. The PBMT bits have been checked for reserved encodings at the source. The expectation is that at the destination - such as PMA checks - the extended PTE bits will intrepreted as PTE_Ext_Flags and then mapped to the symbolic name.

Timmmm commented on model/riscv_vmem.sail:
```diff
@@ -162,9 +166,16 @@ function pt_walk(sv_params,
+                  // Only 64K NAPOT PTEs are valid
+                  PTW_Failure(PTW_Invalid_PTE(), ext_ptw)
+              } else {
+                let final_ppn : bits(64) = match (haveSvnapot(), ext_pte_flags[N]) {
```
I really think an `if` is clearer for these sorts of things. Also you can just reuse the `ppns` name. And use `@` instead of `append`.

```
let ppns = if haveSvnapot() & ext_pte_flags[N] == 0b1
  then ppns[63 .. 4] @ vpn_j[3 .. 0]
  else ppns;
```

ved-rivos replied:
updated.

Timmmm commented on model/riscv_vmem_pte.sail:
```diff
@@ -19,13 +19,20 @@
+bitfield PTE_Ext_Flags : extPte = {
+  N    : 9,
+  PBMT : 8 .. 7,
+  RSVD : 6 .. 0
```
Minor but can we style this like `reserved` so it doesn't look like it is an actual field?

ved-rivos replied:
updated.

Timmmm commented on model/riscv_vmem.sail:
```diff
@@ -275,8 +286,8 @@ function write_pte forall 'n, 'n in {4, 8} . (
 // PUBLIC
-union TR_Result('paddr : Type, 'failure : Type) = {
-  TR_Address : ('paddr, ext_ptw),
+union TR_Result('paddr : Type, 'ext_pte_bits : Type, 'failure : Type) = {
```
`ext_pte_bits` is always `extPte` - is there any point parameterising it?

ved-rivos replied:
please explain more.

Timmmm commented on model/riscv_vmem_pte.sail:
```diff
@@ -51,10 +58,33 @@ function pte_is_ptr(pte_flags : PTE_Flags) -> bool = (pte_flags[X] == 0b0)
                                                       & (pte_flags[R] == 0b0)
 
+// Extension hooks can request standard PTE validity checks by returning Ext_PTE_Std
+union Ext_isInvalPTE_Check = {
```
`enum`

ved-rivos replied:
updated.

Timmmm commented on model/riscv_vmem_pte.sail:
```diff
@@ -51,10 +58,33 @@ function pte_is_ptr(pte_flags : PTE_Flags) -> bool = (pte_flags[X] == 0b0)
+                                 pte_ext_flags[N] != 0b0 | pte_ext_flags[PBMT] != 0b00))
+    }
+  }
+}
```
I think this can be simplified a bit. Here's our version, IIRC I based it on yours.

```
function pte_is_invalid(pte_flags : PTE_Flags, pte_ext : PTE_Ext) -> bool =
  match ext_pte_validity(pte_flags, pte_ext) {
    Ext_PTE_Invalid => true,
    Ext_PTE_Valid   => false,
    Ext_PTE_Std     =>
        pte_flags[V] == 0b0
      | (pte_flags[W] == 0b1 & pte_flags[R] == 0b0)
      // These bits must be zero unless the relevant extension is supported
      // and this is a leaf node.
      | pte_ext[N]    != zeros() & (not(haveSvnapot()) | pte_is_ptr(pte_flags))
      | pte_ext[PBMT] != zeros() & (not(haveSvpbmt())  | pte_is_ptr(pte_flags))
      // PBMT 0b11 is always reserved. This applies even without PBMT support.
      | pte_ext[PBMT] == 0b11
      // Reserved bits must be 0.
      | pte_ext[reserved] != zeros()
  }
```

ved-rivos replied:
This version is missing a few things:
1. A, D, and U flags are reserved in non-leaf PTEs
2. PBMT is reserved if menvcfg.PBMTE is 0
3. N bit is reserved for superpage leaf PTEs

ved-rivos replied:
updated.

ved-rivos replied:
@Timmmm - let me know if any other changes are needed.

pmundkur replied:
@ved-rivos I was taking a look at updating this PR too.  It would be good to split Svnapot and Svpbmt into separate PRs.  Let me know if you're close to completion, or whether I should take a crack at it.

jordancarlin replied:
@pmundkur were you still planning to work on this? Missing Svnapot is starting to become a blocking issue for ACT4 virtual memory tests.

Timmmm reviewed (changes requested):
Looks good to me. There are some minor style issues but easy to fix.

ved-rivos commented on model/riscv_sys_regs.sail:
```diff
@@ -837,6 +849,7 @@ register senvcfg : SEnvcfg
 function legalize_menvcfg(o : MEnvcfg, v : bits(64)) -> MEnvcfg = {
   let v = Mk_MEnvcfg(v);
   let o = [o with FIOM = if sys_enable_writable_fiom() then v[FIOM] else 0b0];
+  let o = [o with PBMTE = if haveSvpbmt() then v[PBMTE] else 0b0];
```
updated.

ved-rivos commented on model/riscv_vmem.sail:
```diff
@@ -151,7 +152,10 @@ function pt_walk(sv_params,
                 let ppns_masked = (ppns >> mask_bits) << mask_bits;
-                if not(ppns == ppns_masked) then
+                if haveSvnapot() & ext_pte_flags[N] == 0b1 then
+                  // Superpage NAPOT PTEs are invalid
```
updated.
